src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/TransactionBundleValidator.cs

```diff
@@ -32,10 +32,12 @@ public TransactionBundleValidator(ResourceReferenceResolver referenceResolver)
         /// <summary>
         /// This method validates if transaction bundle contains multiple entries that are modifying the same resource.
         /// It also validates if the request operations within a entry is a valid operation.
+        /// If a conditional create or update is executed and a resource exists, the value is populated in the idDictionary.
         /// </summary>
         /// <param name="bundle"> The input bundle</param>
+        /// <param name="idDictionary">The id dictionary that stores fullUrl to actual ids.</param>
         /// <param name="cancellationToken"> The cancellation token</param>
-        public async Task ValidateBundle(Hl7.Fhir.Model.Bundle bundle, CancellationToken cancellationToken)
+        public async Task ValidateBundle(Hl7.Fhir.Model.Bundle bundle, IDictionary<string, (string resourceId, string resourceType)> idDictionary, CancellationToken cancellationToken)
```

Comment:
maybe not in this PR, but i feel like this class is doing more and more stuff. It's not just validation anymore. it's also loading the resources, building lookup dictionary.

Perhaps split them out by providing like a pipeline that enumerates over each entries in the bundle?

Reply:
I agree. I think given the time, this could be refactored to increase the readability.